python-lib/prologin/hfs/server.py

```diff
@@ -337,18 +339,33 @@ def remote_user_handler(self, peer_id):
         url = ('http', 'hfs%d:%d' % (peer_id, CFG['port']), '/migrate_user',
                '', '')
         url = urllib.parse.urlunsplit(url)
+
         data = { 'user': self.user, 'hfs': peer_id }
-        data = 'data=%s' % (urllib.parse.quote(json.dumps(data)))
+        data = urllib.parse.urlencode([('data', json.dumps(data))])
         data = data.encode('utf-8')
 
-        with open(self.nbd_filename(), 'wb') as fp:
-            with urllib.request.urlopen(url, data=data) as rfp:
-                with gzip.GzipFile(fileobj=rfp, mode='rb') as zfp:
-                    while True:
-                        block = zfp.read(65536)
-                        if not block:
-                            break
-                        fp.write(block)
+        nbd_path = Path(self.nbd_filename())
+
+        tar_cmdline = ['/usr/bin/tar', '--sparse', '-xzf',
+                       '-', nbd_path.name]
+
+        tar = subprocess.Popen(tar_cmdline,
+                               stdin=subprocess.PIPE,
+                               cwd=nbd_path.parent)
+
+        BLOCK_SIZE = 65536
+
+        with urllib.request.urlopen(url, data=data) as rfp:
+            while True:
+                block = rfp.read(BLOCK_SIZE)
+                if not block:
+                    break
+                tar.stdin.write(block)
+
+        tar.stdin.close()
+
+        if tar.wait() != 0:
+            raise RuntimeError(f"{' '.join(tar_cmdline)} exited with rc")
```

Comment:
Does it crash the RFS server if this happens?

Reply:
nop